### ISSUE TITLE
KafkaAvroDeserializer should not require setting the version id.

### ISSUE THREAD

**Issue by michaelandrepearce** [hortonworks/registry#114]
Currently it seems that the KafkaAvroDeserializer needs to know the version id upfront.

This causes two issues:

Apps that handle messages such as GenericRecord
Apps have the SpecificRecord in the class path/jvm. 

Ideally it should have similar to confluents, where if no READER_VERSION is present, it simply uses the same id as the incoming data. (GenericRecord problem)

If specific record is found, it should find the schema from the specific record found on the class path, it should use its schema

Currently this blocks migration from Confluents Schema Registry.

**Comment by michaelandrepearce** [comment]
Also it seems that if i write with a SpecificRecord, i must read with one, and like wise if i write with a GenericRecord, i must read with one.

This will cause issues, where an app produces with specific and another consumes without the specific (aka some generic integration) and vice-versa.

Should not serial/deserialize to Generic/Specific based on how an array is written, but based on if the app has a SpecificRecord on the CP or not. (similar to confluents)
            
        int recordType = payloadInputStream.read();
                    LOG.debug("Received record type: [{}]", Integer.valueOf(recordType));
                    Object datumReader = null;
                    Schema readerSchema = readerSchemaVersion != null?(Schema)this.getSchema(new SchemaVersionKey(schemaName, readerSchemaVersion)):null;
                    if(recordType == 0) {
                        datumReader = readerSchema != null?new GenericDatumReader(e, readerSchema):new GenericDatumReader(e);
                    } else {
                        datumReader = readerSchema != null?new SpecificDatumReader(e, readerSchema):new SpecificDatumReader(e);
                    }

**Comment by michaelandrepearce** [comment]
https://github.com/hortonworks/registry/pull/115

**Comment by satishd** [comment]
@michaelandrepearce Which version id should be set for KafkaAvroDeserializer?

+1 to be flexible for consuming either specific record or generic record on deserializer side. We can target addressing this issue in upcoming release.

**Comment by michaelandrepearce** [comment]
For what you serialize into for specific it's redundant, as specefic it should take the schema of what's on concrete class on the class path. 

For generic if not provided / present you simply don't evolve and you serialize into generic record using the inbound version. 

For generic if provided then you can evolve as you serialize into generic record 

This is a key point here.

**Comment by michaelandrepearce** [comment]
See the pr, This is all the same logic that's in the confluent serdes also.

**Comment by michaelandrepearce** [comment]
Logic to deal with if no readerSchema (aka no reader version id)

```
+    private DatumReader getDatumReader(Schema writerSchema, Schema readerSchema) {
 +        if(this.useSpecificAvroReader) {
 +            if(readerSchema == null) {
 +                readerSchema = this.getReaderSchema(writerSchema);
 +            }
 +
 +            return new SpecificDatumReader(writerSchema, readerSchema);
 +        } else {
 +            return readerSchema == null?new GenericDatumReader(writerSchema):new GenericDatumReader(writerSchema, readerSchema);
 +        }
 +    }
```

**Comment by satishd** [comment]
Closes #114
